store/store.go

```diff
@@ -770,3 +774,33 @@ type IntegrityCheckResult struct {
 	Data interface{}
 	Err  error
 }
+
+const mySQLDeadlockCode = uint16(1213)
+
+// WithDeadlockRetry retries a given f if it throws a deadlock error.
+// After exceeding a given threshold, it breaks and propagates the error upwards.
+// TODO: This can be a separate retry layer in itself where transaction retries
+// are automatically applied.
+func WithDeadlockRetry(f func() error) error {
+	var err error
+	for i := 0; i < 3; i++ {
+		err = f()
+		if err != nil {
+			// XXX: Possibly add check for postgres deadlocks later.
+			// But deadlocks are very rarely seen in postgres.
+			var mysqlErr *mysql.MySQLError
+			if errors.As(err, &mysqlErr) && mysqlErr.Number == mySQLDeadlockCode {
+				mlog.Warn("A deadlock happened. Retrying.", mlog.Err(err))
+				// This is a deadlock, retry.
+				continue
+			}
+			// Some other error, return as-is.
+			return err
+		} else {
+			// No error, return nil.
+			return nil
+		}
+	}
+	mlog.Error("Deadlock happened 3 times. Giving up")
```

Comment:
This log may be emitted without context, especially if there are multiple things going on in the logs. Should we just wrap the error with a "giving up after 3 deadlocks" or some such? Also, does this deserve to be an `Error`, in that the sysadmin need take some corrective action?

Reply:
> Should we just wrap the error with a "giving up after 3 deadlocks" or some such?

Makes sense.

>  Also, does this deserve to be an Error, in that the sysadmin need take some corrective action?

Yes, I think so. If a deadlock happens consecutively 3 times, there's something seriously wrong somewhere.